### DISCUSSION TITLE
Schemas for Efficient Social Media Queries

### DISCUSSION THREAD

**Discussion by jonaird** [geldata/gel#3105] [Q&A] [answered]
Hi, I'm building a social media type application. When querying for a user's posts I'm wondering whether it would be more efficient to have a user object that has multi-links to all of their posts or to have post objects that link to their user and query for the backlink or whether they should be about the same. Intuitively it sounds like the former would be more efficient. Is that the case?

**Comment by elprans** [comment, marked as answer]
In general, `single` links are more efficient than `multi` links, because a `multi` link involves an intermediate table, which needs to be queried when reading data and updated whenever a new post is created.

Here's a rough guide on when to use `multi` and when to use a backlink, possibly with a computed link on the other end for convenience.

Use a `multi` link if:

1. The relationship is relatively stable, i.e. you don't update it frequently, for example a list of associated postal addresses in a user profile.
2. The number of elements in the link is usually fairly small.

If either of the above isn't true, use a `single` link from the other end and couple it with a computed based on a backlink, like so:

```
type Post {
  required link author -> User;
}

type User {
  link posts := (.<author[IS Post])
}
```

**Reply by jonaird** [reply]
Excellent, thanks!

**Reply by lqhuang** [reply]
Sorry to repeat here @jonaird @elprans . Thanks to your discussion! That helps me a lot. I still have another question. How about those situations must require `multi` links, like `following` and `followers` in social network? I want to build efficient model as possible as I can.

Regards.

**Comment by lqhuang** [comment]
Hey, guys @jonaird @elprans , so many thanks to your discussion! That helps me a lot. I still have another question. How about those situations must require `multi` links, like `following` and `followers` in social network? I want to build efficient model as possible as I can.

Regards.
